torch_geometric/utils/to_dense_batch.py

```diff
@@ -92,17 +93,30 @@ def to_dense_batch(x: Tensor, batch: Optional[Tensor] = None,
     if batch is None:
         batch = x.new_zeros(x.size(0), dtype=torch.long)
 
+    dynamic_shapes_disabled = is_experimental_mode_enabled(
+        "disable_dynamic_shapes")
+
     if batch_size is None:
-        batch_size = int(batch.max()) + 1
+        if dynamic_shapes_disabled:
+            raise RuntimeError(
+                "Dynamic shapes disabled. `batch_size` argument should be set."
+            )
+        else:
+            batch_size = int(batch.max()) + 1
 
     num_nodes = scatter(batch.new_ones(x.size(0)), batch, dim=0,
                         dim_size=batch_size, reduce='sum')
     cum_nodes = torch.cat([batch.new_zeros(1), num_nodes.cumsum(dim=0)])
 
     filter_nodes = False
     if max_num_nodes is None:
-        max_num_nodes = int(num_nodes.max())
-    elif num_nodes.max() > max_num_nodes:
+        if dynamic_shapes_disabled:
+            raise RuntimeError(
+                "Dynamic shapes disabled. `max_num_nodes` argument should be "
+                "set.")
+        else:
+            max_num_nodes = int(num_nodes.max())
+    elif not dynamic_shapes_disabled and num_nodes.max() > max_num_nodes:
```

Comment:
I am wondering if we can make this a decorator instead? For example,
```python
@disable_dynamic_shapes(required_args=['batch_size', 'max_num_nodes'])
def from_dense_batch(...):
    ...
```
seems a bit cleaner than adding this logic to all functions manually.

Reply:
The idea with the decorator looks nice, especially for the validation of the parameters. But what about the possibility of conditional blocking some piece of code or replacing the implementation ? For example, in 119 I want to disable the execution of that part of the code.

Reply:
Is there a particular reason you want to disable this code-branch here at all? I am okay with only using the decorator for basic validation, and then use `dynamic_shapes_disabled` for some more advanced custom logic (although I am still not sure it is needed here).

Reply:
I'm sorry for not getting back to you sooner. I was on sick leave.
1. I'm okay with the decorator solution. I will prepare a patch with that.
2. I want to disable that part of code because filtering_nodes produces a dynamic mask, which can produce a dynamic shape on the x. Disabling dynamic shapes in torch dynamo will produce an error here during torch compile, it is also a problem for static compiler backends like ours (GC).

Reply:
@rusty1s I pushed change with a decorator.